### PR TITLE
369 allowing tests to be performed at individual locations not only location types

### LINKED ISSUE #369: Allowing tests to be performed at individual locations, not only location types
**Is your feature request related to a problem? Please describe**
Related to #326
To allow for more possibilities, testing should not only be possible in none or all location types, but also for individual locations.

**Describe the solution you'd like**
If possible, a more general, templated solution for TestingCriteria could be implemented where lot's of flexibility is given.

**Describe alternatives you've considered**
Alternatively, a similar class can be added which implements this feature.

**Additional context**
This should be done in a similar way for NPIs.

### LINKED ISSUE #755: Simplify and optimize TestingScheme
### Motivation / Current Behaviour

The TestingScheme is currently one of the most expensive parts of the main simulation loop. There is also a lot of dynamic data (vectors). There are significant performance gains possible.

### Enhancement description

- Use bitset (one bit per location type/infection state/age) instead of vector for TestingCriteria. There is a fixed number of location types, infection states, etc. No need for a dynamic container.
- Only one TestingCriteria per TestingScheme. Removes more dynamic data.
- We should also think about not running all TestingSchemes in all cases, check location first maybe?

Each of the first two changes is ~5% runtime. The first change is always good. the gain for the second change depends on how often there are testingschemes with multiple criterias. Right now there is always only one Criteria anyway. If this leads to many more TestingSchemes, it may not improve performance. But the code is still much simpler.

The third change needs to be discussed. But the potential gain is huge, not running the schemes or only running them conditionally potentially decreases runtime by 50%.

An almost finished implementation of the first two changes is here: 629-parallelize-abm-optimize-testing

### Additional context

_No response_

### Checklist

- [X] Attached labels, especially loc:: or model:: labels.
- [X] Linked to project

### PR DESCRIPTION
**Pull request by khoanguyen-dev**
# Changes and Information

Please **briefly list the changes** made, additional Information and what the Reviewer should look out for:

- Remove Location from TestingCriteria
- Use bitset for infectionState and unordered_set for AgeGroup instead of vector for TestingCriteria
- Only one TestingCriteria per TestingScheme
- Add map of LocationID to TestingScheme in TestingStrategy so run_strategy() will check Location and run only the corresponding schemes. 

## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/DLR-SC/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/DLR-SC/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added

### Checks by code reviewer(s)

- [ ] Corresponding issue(s) is/are linked and addressed
- [ ] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [ ] Appropriate **unit tests** have been added, CI passes and code coverage is acceptable (did not decrease)
- [ ] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)

Closes #369
Closes #755

### CODE REVIEW

**Review by xsaschako — CHANGES_REQUESTED** [review]
looks good, a few quesions and comments remain

**Review by xsaschako — CHANGES_REQUESTED** [review]
looks good, a few quesions and comments remain

**Review by xsaschako — CHANGES_REQUESTED** [review]
looks good, a few quesions and comments remain

File: cpp/models/abm/testing_strategy.cpp [hunk 1]
```diff
@@ -26,102 +26,49 @@ namespace mio
...
 
-bool TestingCriteria::has_requested_infection_state(const Person& p, TimePoint t) const
+bool TestingCriteria::evaluate(const Person& p, TimePoint t) const
 {
-    if (m_infection_states.empty()) {
-        return true; // no condition on infection state
-    }
-    return std::find(m_infection_states.begin(), m_infection_states.end(), p.get_infection_state(t)) !=
-           m_infection_states.end();
+    return (m_ages.empty() || m_ages.count(static_cast<size_t>(p.get_age()))) &&
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.cpp:67]: Yes, I added some comments there now.

File: cpp/models/abm/testing_strategy.cpp [hunk 2]
```diff
@@ -142,84 +89,92 @@ bool TestingScheme::operator==(const TestingScheme& other) const
...
+    : m_location_to_schemes_map(location_to_schemes_map)
 {
 }
 
-void TestingStrategy::add_testing_scheme(const TestingScheme& scheme)
+void TestingStrategy::add_testing_scheme(const LocationId& loc_id, const TestingScheme& scheme)
 {
-    if (std::find(m_testing_schemes.begin(), m_testing_schemes.end(), scheme) == m_testing_schemes.end()) {
-        m_testing_schemes.push_back(scheme);
+    auto& schemes_vector = m_location_to_schemes_map[loc_id];
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.cpp:124]: I think it's a vector since the map is from LocationId to a vector of TestingScheme.

File: cpp/models/abm/testing_strategy.h [hunk 3]
```diff
@@ -94,35 +84,14 @@ class TestingCriteria
...
-     * @param[in] p Person to be checked.
-     * @param[in] t TimePoint when to check.
-     */
-    bool has_requested_infection_state(const Person& p, TimePoint t) const;
-
-    std::vector<AgeGroup> m_ages; ///< Set of #AgeGroup%s that are either allowed or required to be tested.
-    std::vector<LocationType> m_location_types; /**< Set of #LocationState%s that are either allowed or required to be 
-    tested.*/
-    std::vector<InfectionState> m_infection_states; /**< Set of #InfectionState%s that are either allowed or required to
+    std::unordered_set<size_t> m_ages; ///< Set of #AgeGroup%s that are either allowed or required to be tested.
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.h:94]: Bitset was suggested by Daniel to avoid using vectors for both AgeGroup and InfectionState.
However, now the number of AgeGroup is not defined until runtime so bitset is not possible.

File: cpp/models/abm/testing_strategy.h [hunk 4]
```diff
@@ -178,14 +135,13 @@ class TestingScheme
...
      * @param[in] person Person to check.
-     * @param[in] location Location to check.
      * @param[in] t TimePoint when to run the scheme.
      * @return If the person is allowed to enter the Location by the scheme.
      */
-    bool run_scheme(Person::RandomNumberGenerator& rng, Person& person, const Location& location, TimePoint t) const;
+    bool run_scheme(Person::RandomNumberGenerator& rng, Person& person, TimePoint t) const;
 
 private:
-    std::vector<TestingCriteria> m_testing_criteria; ///< Vector with all TestingCriteria of the scheme.
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.h:188]: This one was suggested by Daniel because we don't have any TestingScheme with multiple TestingCriteria in the simulation (only in tests). In this way, we avoid using vectors in exchange for more schemes to be defined.

File: cpp/models/abm/testing_strategy.cpp [hunk 5]
```diff
@@ -142,84 +89,92 @@ bool TestingScheme::operator==(const TestingScheme& other) const
...
+    auto last            = std::remove(schemes_vector.begin(), schemes_vector.end(), scheme);
+    schemes_vector.erase(last, schemes_vector.end());
+}
+
+void TestingStrategy::remove_testing_scheme(const LocationType& loc_type, const TestingScheme& scheme)
 {
-    auto last = std::remove(m_testing_schemes.begin(), m_testing_schemes.end(), scheme);
-    m_testing_schemes.erase(last, m_testing_schemes.end());
+    auto loc_id = LocationId{INVALID_LOCATION_INDEX, loc_type};
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.cpp]: It's adjusted as suggested.

File: cpp/models/abm/testing_strategy.cpp [hunk 6]
```diff
@@ -142,84 +89,92 @@ bool TestingScheme::operator==(const TestingScheme& other) const
...
+        schemes_vector.emplace_back(scheme);
     }
 }
 
-void TestingStrategy::remove_testing_scheme(const TestingScheme& scheme)
+void TestingStrategy::add_testing_scheme(const LocationType& loc_type, const TestingScheme& scheme)
+{
+    auto loc_id = LocationId{INVALID_LOCATION_INDEX, loc_type};
+    add_testing_scheme(loc_id, scheme);
+}
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.cpp]: It's adjusted as suggested.

File: cpp/models/abm/testing_strategy.h [hunk 7]
```diff
@@ -40,19 +42,19 @@ class TestingCriteria
     /**
      * @brief Create a TestingCriteria.
      * @param[in] ages Vector of AgeGroup%s that are either allowed or required to be tested.
-     * @param[in] location_types Vector of #LocationType%s that are either allowed or required to be tested.
+     * @param[in] locations Vector of #Location%s or #LocationType%s that are either allowed or required to be tested.
      * @param[in] infection_states Vector of #InfectionState%s that are either allowed or required to be tested.
      * An empty vector of ages/#LocationType%s/#InfectionStates% means that no condition on the corresponding property
      * is set!
      */
     TestingCriteria() = default;
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.h]: Yes, I added a comment to make it clearer.

File: cpp/models/abm/testing_strategy.h [hunk 8]
```diff
@@ -178,14 +135,13 @@ class TestingScheme
...
-     * @param[in] location Location to check.
      * @param[in] t TimePoint when to run the scheme.
      * @return If the person is allowed to enter the Location by the scheme.
      */
-    bool run_scheme(Person::RandomNumberGenerator& rng, Person& person, const Location& location, TimePoint t) const;
+    bool run_scheme(Person::RandomNumberGenerator& rng, Person& person, TimePoint t) const;
 
 private:
-    std::vector<TestingCriteria> m_testing_criteria; ///< Vector with all TestingCriteria of the scheme.
+    TestingCriteria m_testing_criteria; ///< Vector with all TestingCriteria of the scheme.
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.h]: It is now updated.

File: cpp/models/abm/testing_strategy.h [hunk 9]
```diff
@@ -40,19 +42,19 @@ class TestingCriteria
     /**
      * @brief Create a TestingCriteria.
      * @param[in] ages Vector of AgeGroup%s that are either allowed or required to be tested.
-     * @param[in] location_types Vector of #LocationType%s that are either allowed or required to be tested.
+     * @param[in] locations Vector of #Location%s or #LocationType%s that are either allowed or required to be tested.
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/testing_strategy.h]: Yes, it is removed from the comment.

File: cpp/models/abm/location_type.h [hunk 10]
```diff
@@ -90,4 +99,15 @@ struct GeographicalLocation {
 } // namespace abm
 } // namespace mio
 
+namespace std
```
[the remark this answers is not in the document]
Comment (khoanguyen-dev) [reply cpp/models/abm/location_type.h]: Yes, I adjusted accordingly.

**Review by xsaschako — APPROVED** [review]
looks good! Thank you
